tests/test_geospatial.py

```diff
@@ -0,0 +1,418 @@
+import glob
+import json
+import os
+import shutil
+import tempfile
+import unittest
+import warnings
+from pathlib import Path
+from unittest import mock
+
+import affine
+import fiona
+import numpy as np
+import rasterio
+import shapely
+
+import tiledb
+from tiledb.cloud.geospatial import BoundingBox
+from tiledb.cloud.geospatial import DatasetType
+from tiledb.cloud.geospatial import build_inputs_udf
+from tiledb.cloud.geospatial import ingest_geometry_udf
+from tiledb.cloud.geospatial import ingest_point_cloud_udf
+from tiledb.cloud.geospatial import ingest_raster_udf
+from tiledb.cloud.geospatial import load_geometry_metadata
+from tiledb.cloud.geospatial import load_pointcloud_metadata
+from tiledb.cloud.geospatial import load_raster_metadata
+from tiledb.cloud.geospatial import read_uris
+from tiledb.cloud.utilities import chunk
+from tiledb.cloud.utilities import serialize_filter
+from tiledb.vfs import VFS
+
+RASTER_NAMES = [
+    "test1.tif",
+    "test2.tif",
+    "test3.tif",
+    "test_diff_res.tif",
+    "test_diff_crs.tif",
+]
+
+PC_NAMES = glob.glob(
+    f"{os.path.join(os.path.dirname(__file__), 'data', 'geospatial')}/*.las"
+)
+
+GEOM_NAMES = ["test1.geojson", "test2.geojson", "test3.geojson"]
+
+
+def run_local(
+    dataset_uri: str, *, dataset_type: DatasetType, batch_size: int, **kwargs
+):
+    # run the key functions of the DAG locally in sequence
+    funcs = {
+        DatasetType.POINTCLOUD: {
+            "udf_fn": ingest_point_cloud_udf,
+        },
+        DatasetType.RASTER: {
+            "udf_fn": ingest_raster_udf,
+        },
+        DatasetType.GEOMETRY: {
+            "udf_fn": ingest_geometry_udf,
+        },
+    }
+
+    fn = funcs[dataset_type]["udf_fn"]
+
+    # unit test inner functions
+    new_kwargs = build_inputs_udf(
+        dataset_type=dataset_type,
+        **kwargs,
+    )
+
+    # set up function
+    samples = fn(
+        dataset_uri=dataset_uri,
+        append=False,
+        batch_size=batch_size,
+        **new_kwargs,
+    )
+
+    new_kwargs.pop("sources")
+
+    # work functions
+    for _, work in enumerate(samples):
+        fn(
+            sources=work,
+            dataset_uri=dataset_uri,
+            append=True,
+            **new_kwargs,
+        )
+
+
+class GeospatialTest(unittest.TestCase):
+    def setUp(self):
+        # Ignore warnings
+        warnings.simplefilter("ignore")
+        # Create a temporary directory
+        self.test_dir = Path(tempfile.mkdtemp())
+
+        def create_test_geometries(tmp_path: os.PathLike):
+            radius = 1.0
+            for i in range(1, 4):
+                g = shapely.Point(i, i).buffer(radius)
+                with open(tmp_path.joinpath(GEOM_NAMES[i - 1]), "w") as dst:
+                    json.dump(shapely.geometry.mapping(g), dst)
+
+        def create_test_rasters(tmp_path: os.PathLike):
+            kwargs = {
```

Comment:
Likewise, at the top of this function, do `import affine; import rasterio`, etc.

This means that when this file is imported by `pytest`, it won’t immediately import any of the libraries, and by default all these tests will be skipped.